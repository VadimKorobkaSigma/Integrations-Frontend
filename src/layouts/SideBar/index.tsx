import * as React from 'react';
import cn from 'classnames';
import SVG from 'react-inlinesvg';
import logoIcon from '@assets/images/logo.png';
import settingsIcon from '@assets/images/threeDots.svg';
import orgIcon from '@assets/images/orgIcon.svg';
import styles from './styles.module.scss';
import SettingsModal from '@layouts/SettingsModal';
import Organization from '@dtos/organization';
import ErrorComponent from '@components/ErrorComponent';
import useSearch from '@services/hooks/useSearch';
import Search, { SearchHighlight } from '@components/Search';

interface Props {
    organizations: Organization[];
    selectedOrg: string;
    error: string;
    isLoading: boolean;
    setSelectedOrg: React.Dispatch<React.SetStateAction<string>>;
}

const SideBar: React.FC<Props> = ({ error, organizations, selectedOrg, setSelectedOrg, isLoading }) => {
    const [settingsForOrg, setSettingsForOrg] = React.useState<Organization | null>(null);
    const [filteredOrgs, searchValue, setSearchValue] = useSearch(organizations, 'name');
    const selectOrg = (e: React.MouseEvent<HTMLLIElement>) => {
        if (e.target instanceof Element) {
            const id = e.target.getAttribute('data-name') || e.target.parentElement?.getAttribute('data-name');
            if (id) {
                setSelectedOrg(id);
            }
        }
    };

    const openSettings = (e: React.MouseEvent<SVGElement>, org: Organization) => {
        e.stopPropagation();
        setSettingsForOrg(org);
    };

    return (
        <section className={cn(styles.column, styles.sidebar)}>
            <div className={styles.header}>
                <img src={logoIcon} alt="logo" />
                <h2>Organizations</h2>
            </div>
            <ErrorComponent error={error} />
            {isLoading && <h3 className={styles.message}>Loading...</h3>}
            {!isLoading && organizations.length ? (
                <>
                    <Search containerClassname={styles.search} value={searchValue} onChange={setSearchValue} />
                    <ul>
                        {filteredOrgs.map((org) => (
                            <li
                                key={org.id}
                                data-name={org.id}
                                onClick={selectOrg}
                                className={cn(styles.item, selectedOrg === org.id && styles.active)}
                            >
                                <SVG src={orgIcon} />
                                <SearchHighlight searchQuery={searchValue} text={org.name} />
                                <SVG
                                    src={settingsIcon}
                                    data-name={org.id}
                                    className={styles.settings}
                                    onClick={(e) => openSettings(e, org)}
                                />
                            </li>
                        ))}
                    </ul>
                </>
            ) : null}
            <SettingsModal selectedOrg={settingsForOrg} setSelectedOrg={setSettingsForOrg} />
        </section>
    );
};

export default React.memo(SideBar);
