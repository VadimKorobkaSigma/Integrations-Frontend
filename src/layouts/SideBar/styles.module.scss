.column {
  display: flex;
  flex-direction: column;
}

.header {
  display: flex;

  img {
    width: 24px;
    height: 24px;
    margin-right: 5px;
  }
}

.sidebar {
  background-color: #f6f6f6;
  padding: 20px;

  .message {
    color: $text-black;
    margin: 20px 0;
  }

  .item {
    display: flex;
    height: 60px;
    cursor: pointer;

    span {
      text-overflow: ellipsis;
      overflow: hidden;
      white-space: nowrap;
      font-size: 18px;
      line-height: 60px;
      letter-spacing: -0.005em;
      color: $text-black;
    }

    &.active {
      border-right: 4px solid $green;
      font-weight: 500;

      .settings {
        margin-right: -4px;
      }
    }

    * {
      cursor: pointer;
    }

    .settings {
      margin-left: auto;
    }
  }
}
