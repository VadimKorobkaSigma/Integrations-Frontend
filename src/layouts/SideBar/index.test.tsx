import * as React from 'react';
import SideBar from './index';
import SVG from 'react-inlinesvg';
import { shallow } from 'enzyme';
import OrgSettings from '@dtos/orgSettings';

const data = [
    { id: 'test_id', name: 'test 1' },
    { id: 'test_id_2', name: 'test 2' },
];

test('SideBar Component', () => {
    const selectedOrg = 'test_id_2';
    const setSelectedOrg = jest.fn();
    const setSettingsForOrg = jest.fn();
    global.fetch = jest.fn().mockImplementation(() =>
        Promise.resolve({
            ok: true,
            json: () => Promise.resolve({ cxgoToken: 'cxgoToken', team: 'team' } as OrgSettings),
        }),
    );
    jest.spyOn(React, 'useState').mockImplementation(
        jest
            .fn()
            .mockReturnValueOnce([selectedOrg, setSettingsForOrg])
            .mockReturnValueOnce(['', jest.fn()])
            .mockReturnValueOnce([data, jest.fn()])
            .mockReturnValue([null, jest.fn()]),
    );

    const wrapper = shallow(
        <SideBar
            organizations={data}
            selectedOrg={selectedOrg}
            error=""
            isLoading={false}
            setSelectedOrg={setSelectedOrg}
        />,
    );
    expect(wrapper.html()).toMatchSnapshot();
    wrapper.find(SVG).last().simulate('click', new MouseEvent('click'));
    wrapper.find('li').last().simulate('click', new MouseEvent('click'));
});

test('SideBar Component', () => {
    const selectedOrg = 'test_id_2';
    const setSelectedOrg = jest.fn();
    const setSettingsForOrg = jest.fn();
    global.fetch = jest.fn().mockImplementation(() =>
        Promise.resolve({
            ok: true,
            json: () => Promise.resolve({ cxgoToken: 'cxgoToken', team: 'team' } as OrgSettings),
        }),
    );
    jest.spyOn(React, 'useState').mockImplementation(
        jest
            .fn()
            .mockReturnValueOnce([selectedOrg, setSettingsForOrg])
            .mockReturnValueOnce(['', jest.fn()])
            .mockReturnValueOnce([data, jest.fn()])
            .mockReturnValue([null, jest.fn()]),
    );

    const wrapper = shallow(
        <SideBar organizations={[]} selectedOrg={''} error="" isLoading={true} setSelectedOrg={setSelectedOrg} />,
    );
    expect(wrapper.html()).toMatchSnapshot();
});
