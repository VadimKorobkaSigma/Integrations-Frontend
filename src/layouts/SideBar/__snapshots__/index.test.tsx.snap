// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideBar Component 1`] = `"<section class=\\"\\"><div><img alt=\\"logo\\"/><h2>Organizations</h2></div><div class=\\"\\"><input placeholder=\\"Search\\" value=\\"\\" class=\\"\\"/></div><ul><li data-name=\\"test_id\\" class=\\"\\"><span>test 1</span></li><li data-name=\\"test_id_2\\" class=\\"\\"><span>test 2</span></li></ul><div class=\\"\\"><div></div><div><header><h1>Settings of undefined organization</h1></header><form><label for=\\"secret\\">Checkmarx Go Secret</label><textarea placeholder=\\"Secret\\" id=\\"secret\\"></textarea><label for=\\"team\\">Team</label><textarea placeholder=\\"Team\\" id=\\"team\\"></textarea></form><footer><button variant=\\"link\\" class=\\"\\">Cancel</button><button class=\\"\\">Submit</button></footer></div></div></section>"`;

exports[`SideBar Component 2`] = `"<section class=\\"\\"><div><img alt=\\"logo\\"/><h2>Organizations</h2></div><h3>Loading...</h3><div class=\\"\\"><div></div><div><header><h1>Settings of undefined organization</h1></header><form><label for=\\"secret\\">Checkmarx Go Secret</label><textarea placeholder=\\"Secret\\" id=\\"secret\\"></textarea><label for=\\"team\\">Team</label><textarea placeholder=\\"Team\\" id=\\"team\\"></textarea></form><footer><button variant=\\"link\\" class=\\"\\">Cancel</button><button class=\\"\\">Submit</button></footer></div></div></section>"`;
